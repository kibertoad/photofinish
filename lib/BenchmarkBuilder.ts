import {
  BenchmarkResults,
  executeBenchmarkAsync,
  executeBenchmarkSync,
} from './benchmarkExecutioner'

export type FunctionType = () => any
export type AsyncFunctionType = () => Promise<any>

export type Benchmark = {
  name: string
  warmupCycles: number // How many times function will be run before benchmarking starts in order to allow Node.js to perform optimizations
  benchmarkCycles: number // How many execution cycles are executed in order to collect samples
  benchmarkCycleSamples: number // How many samples per cycle are generated. Note that samples below 15 and above 85 percentile are discarded as insignificant
  functionUnderTest: FunctionType | undefined
  asyncFunctionUnderTest: AsyncFunctionType | undefined
  execute: () => BenchmarkResults
  executeAsync: () => Promise<BenchmarkResults>
}

export class BenchmarkBuilder {
  private _name = 'Benchmark'
  private _warmupCycles = 1000
  private _benchmarkCycles = 20000
  private _benchmarkCycleSamples = 100
  private _functionUnderTest: (() => any) | undefined
  private _asyncFunctionUnderTest: (() => Promise<any>) | undefined

  name(name: string): BenchmarkBuilder {
    this._name = name
    return this
  }

  warmupCycles(warmupCycles: number): BenchmarkBuilder {
    this._warmupCycles = warmupCycles
    return this
  }

  benchmarkCycles(benchmarkCycles: number): BenchmarkBuilder {
    this._benchmarkCycles = benchmarkCycles
    return this
  }

  benchmarkCycleSamples(benchmarkCycleSamples: number): BenchmarkBuilder {
    this._benchmarkCycleSamples = benchmarkCycleSamples
    return this
  }

  functionUnderTest(
    functionUnderTest: FunctionType,
    isPromiseResultAllowed = false
  ): BenchmarkBuilder {
    if (this._asyncFunctionUnderTest) {
      throw new Error(
        'asyncFunctionUnderTest is already set, please only set one of asyncFunctionUnderTest and functionUnderTest'
      )
    }

    const sanityCheckResult = functionUnderTest()
    if (!isPromiseResultAllowed && isPromise(sanityCheckResult)) {
      throw new Error(
        `Please use .asyncFunctionUnderTest() method for benchmarking async functions. If you are sure about what you are doing and don't want to await returned promise within benchmark, please set "isPromiseResultAllowed" parameter to true`
      )
    }

    this._functionUnderTest = functionUnderTest
    return this
  }

  asyncFunctionUnderTest(asyncFunctionUnderTest: AsyncFunctionType): BenchmarkBuilder {
    if (this._functionUnderTest) {
      throw new Error(
        'functionUnderTest is already set, please only set one of asyncFunctionUnderTest and functionUnderTest'
      )
    }

    this._asyncFunctionUnderTest = asyncFunctionUnderTest
    return this
  }

  build(): Benchmark {
    if (!this._asyncFunctionUnderTest && !this._functionUnderTest) {
      throw new Error('Either asyncFunctionUnderTest or functionUnderTest needs to bet set.')
    }

    const benchmark: Benchmark = {
      name: this._name,
      warmupCycles: this._warmupCycles,
      benchmarkCycles: this._benchmarkCycles,
      benchmarkCycleSamples: this._benchmarkCycleSamples,
      asyncFunctionUnderTest: this._asyncFunctionUnderTest,
      functionUnderTest: this._functionUnderTest,
      execute: () => {
        throw new Error('Function under test is asynchronous, use executeAsync() instead.')
      },
      executeAsync: () => {
        throw new Error('Function under test is synchronous, use execute() instead.')
      },
    }

    if (benchmark.functionUnderTest) {
      benchmark.execute = (): BenchmarkResults => {
        return executeBenchmarkSync(benchmark)
      }
    }

    if (benchmark.asyncFunctionUnderTest) {
      benchmark.executeAsync = (): Promise<BenchmarkResults> => {
        return executeBenchmarkAsync(benchmark)
      }
    }

    return benchmark
  }
}

function isPromise(value: any) {
  return Promise.resolve(value) === value
}
