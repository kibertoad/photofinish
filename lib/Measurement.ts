export class Measurement {
  private readonly timeInNanoSeconds: number
  constructor(timeInNanoSeconds: number) {
    this.timeInNanoSeconds = timeInNanoSeconds
  }

  getTimeInNanoSeconds(): number {
    return this.timeInNanoSeconds
  }

  getTimeInMilliSeconds(): number {
    return this.timeInNanoSeconds / 1000000
  }

  getTextInMsecs(): string {
    return `${this.getTimeInMilliSeconds().toFixed(3)} msecs`
  }

  getOpsPerSec(): number {
    return 1000000000 / this.getTimeInNanoSeconds()
  }

  getTextOpsPerSec(): string {
    return this.getOpsPerSec().toFixed(3)
  }
}
