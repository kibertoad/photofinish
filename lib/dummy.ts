export function dummy(): void {
  console.log('hi')
}
