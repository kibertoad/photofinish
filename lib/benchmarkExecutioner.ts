import { Benchmark, FunctionType } from './BenchmarkBuilder'

export type BenchmarkResults = {
  avgTime: number
}

async function warmupAsync(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    await functionUnderTest()
  }
}

function warmup(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    functionUnderTest()
  }
}

export function executeBenchmarkSync(benchmark: Benchmark): BenchmarkResults {
  if (!benchmark.functionUnderTest) {
    throw new Error('Function under test is not set')
  }
  warmup(benchmark.warmupCycles, benchmark.functionUnderTest)

  return {
    avgTime: 0,
  }
}

export async function executeBenchmarkAsync(benchmark: Benchmark): Promise<BenchmarkResults> {
  if (!benchmark.asyncFunctionUnderTest) {
    throw new Error('Async function under test is not set')
  }
  await warmupAsync(benchmark.warmupCycles, benchmark.asyncFunctionUnderTest)

  return {
    avgTime: 0,
  }
}
