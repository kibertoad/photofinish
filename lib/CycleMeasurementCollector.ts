import { Measurement } from './Measurement'

export type CycleMeasurements = {
  samples: number[]

  mean: Measurement
}

export class CycleMeasurementCollector {}
