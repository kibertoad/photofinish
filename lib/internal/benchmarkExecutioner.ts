import { AsyncFunctionType, Benchmark, FunctionType } from '../BenchmarkBuilder'
import { Measurement } from '../Measurement'
import { filterSignificantSamples, calculateMean } from './mathUtils'

export type BenchmarkResults = {
  benchmarkName: string
  benchmarkEntryName: string
  meanTime: Measurement
}

async function warmupAsync(warmupCycles: number, functionUnderTest: AsyncFunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    await functionUnderTest()
  }
}

function warmup(warmupCycles: number, functionUnderTest: FunctionType) {
  for (let i = 0; i < warmupCycles; i++) {
    functionUnderTest()
  }
}

export function executeBenchmarkSync(benchmark: Benchmark): BenchmarkResults {
  if (!benchmark.functionUnderTest) {
    throw new Error('Function under test is not set')
  }
  warmup(benchmark.warmupCycles, benchmark.functionUnderTest)

  // Perform all cycles
  const significantSamples = []
  for (let i = 0; i < benchmark.benchmarkCycles; i++) {
    // Perform full cycle
    const samples = []
    for (let i = 0; i < benchmark.benchmarkCycleSamples; i++) {
      const hrStart = process.hrtime()
      benchmark.functionUnderTest()
      const hrEnd = process.hrtime(hrStart)
      const [, timeTakenInNanoSeconds] = hrEnd
      samples.push(timeTakenInNanoSeconds)
    }
    const significantCycleSamples = filterSignificantSamples(samples)
    significantSamples.push(...significantCycleSamples)
  }

  const meanTime = calculateMean(significantSamples)

  return {
    benchmarkName: benchmark.benchmarkName,
    benchmarkEntryName: benchmark.benchmarkEntryName,
    meanTime: new Measurement(meanTime),
  }
}

export async function executeBenchmarkAsync(benchmark: Benchmark): Promise<BenchmarkResults> {
  if (!benchmark.asyncFunctionUnderTest) {
    throw new Error('Async function under test is not set')
  }
  await warmupAsync(benchmark.warmupCycles, benchmark.asyncFunctionUnderTest)

  // Perform all cycles
  const significantSamples = []
  for (let i = 0; i < benchmark.benchmarkCycles; i++) {
    // Perform full cycle
    const samples = []
    for (let i = 0; i < benchmark.benchmarkCycleSamples; i++) {
      const hrStart = process.hrtime()
      await benchmark.asyncFunctionUnderTest()
      const hrEnd = process.hrtime(hrStart)
      const [, timeTakenInNanoSeconds] = hrEnd
      samples.push(timeTakenInNanoSeconds)
    }
    const significantCycleSamples = filterSignificantSamples(samples)
    significantSamples.push(...significantCycleSamples)
  }

  const meanTime = calculateMean(significantSamples)

  return {
    benchmarkName: benchmark.benchmarkName,
    benchmarkEntryName: benchmark.benchmarkEntryName,
    meanTime: new Measurement(meanTime),
  }
}
