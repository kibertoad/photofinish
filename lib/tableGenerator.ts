import { TableData } from './resultsExporter'
import { Measurement } from './Measurement'

const table = require('markdown-table')

export type TableOptions = {
  precision?: number
}

export function generateTable(results: readonly TableData[], options: TableOptions = {}): string {
  const tableIndex: string[] = ['Node', 'Option', 'Msecs/op', 'Ops/sec', 'V8']

  const tableRows: string[][] = [tableIndex]
  results.map((entry) => {
    const measurement = new Measurement(entry.meanTimeNs)
    const [nodeVersion, v8Version] = entry.runtimeVersion.split(', ')

    const row = [
      nodeVersion,
      entry.benchmarkEntryVersion
        ? `${entry.benchmarkEntryName} ${entry.benchmarkEntryVersion}`
        : entry.benchmarkEntryName,
      measurement.getTextInMsecs(options.precision ?? 3),
      measurement.getTextOpsPerSec(),
      v8Version,
    ]

    tableRows.push(row)
  })

  return table(tableRows)
}
