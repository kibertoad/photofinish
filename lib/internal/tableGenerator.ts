const table = require('markdown-table')

import { Benchmark } from '../BenchmarkBuilder'
import { BenchmarkResults } from './benchmarkExecutioner'

export function generateTable(
  benchmark: Benchmark,
  results: readonly BenchmarkResults[]
): string[][] {
  const tableRows: string[][] = [['Option', 'Msecs/op', 'Ops/sec']]
  results.map((entry) => {
    tableRows.push([
      entry.benchmarkEntryName,
      entry.meanTime.getTextInMsecs(),
      entry.meanTime.getTextOpsPerSec(),
    ])
  })

  return table(tableRows)
}
