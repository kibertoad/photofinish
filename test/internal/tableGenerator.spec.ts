import { BenchmarkBuilder } from '../../lib/BenchmarkBuilder'
import { Measurement } from '../../lib/Measurement'
import { BenchmarkResults } from '../../lib/internal/benchmarkExecutioner'
import { generateTable } from '../../lib/internal/tableGenerator'

function sleep(ms: number): Promise<any> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

describe('tableGenerator', () => {
  describe('generateTable', () => {
    it('generate correct table', async () => {
      const benchmark = new BenchmarkBuilder()
        .benchmarkName('Dummy benchmark')
        .warmupCycles(1)
        .benchmarkCycles(10)
        .benchmarkCycleSamples(1)
        .asyncFunctionUnderTest(() => sleep(100))
        .build()

      const result: BenchmarkResults = {
        benchmarkName: 'Dummy benchmark',
        benchmarkEntryName: 'Benchmark entry',
        meanTime: new Measurement(101744870),
      }

      expect(generateTable(benchmark, [result])).toMatchSnapshot()
    })
  })
})
