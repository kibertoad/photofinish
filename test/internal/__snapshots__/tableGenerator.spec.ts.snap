// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tableGenerator generateTable generate correct table 1`] = `
"| Option          | Msecs/op      | Ops/sec |
| --------------- | ------------- | ------- |
| Benchmark entry | 101.745 msecs | 9.829   |"
`;
